receiver/prometheusexecreceiver/receiver.go

```diff
@@ -0,0 +1,295 @@
+// Finally, the wait time before the subprocess is restarted is computed and this goroutine sleeps for that amount of time, before restarting the loop from the start
+func (wrapper *prometheusExecReceiver) manageProcess(ctx context.Context, host component.Host) {
+	var (
+		newPort int = wrapper.originalPort
+
+		elapsed       time.Duration
+		crashCount    int
+		err           error
+		subprocessErr error
+	)
+
+	for {
+
+		// Generate a port if none was specified and if process had a low runtime (Receiver was stopped)
+		generatePort := wrapper.originalPort == 0 && elapsed <= healthyProcessTime
+		if generatePort {
+			newPort, err = wrapper.assignNewRandomPort(ctx)
+			if err != nil {
+				wrapper.logger.Error("assignNewRandomPort() error - killing this single process/receiver", zap.String("error", err.Error()))
+				return
+			}
+		}
+
+		wrapper.subprocessConfig = wrapper.fillPortPlaceholders(newPort)
+
+		// Start the receiver if it's the first pass, or if the process is unhealthy/a newport was generated meaning the Receiver was stopped last pass
+		firstRun := elapsed == 0
+		unhealthyProcess := elapsed <= healthyProcessTime && crashCount > healthyCrashCount
+
+		if firstRun || unhealthyProcess || generatePort {
+			err = wrapper.prometheusReceiver.Start(ctx, host)
+			if err != nil {
+				wrapper.logger.Error("Start() error, could not start receiver - killing this single process/receiver", zap.String("error", err.Error()))
+				return
+			}
+		}
+
+		// Start the process, error is handled a little later
+		elapsed, subprocessErr = wrapper.subprocessConfig.Run(wrapper.logger)
+
+		crashCount, err = wrapper.computeHealthAndCrashCount(ctx, elapsed, crashCount)
+		if err != nil {
+			wrapper.logger.Error("computeHealthAndCrashCount() error - killing this single process/receiver", zap.String("error", err.Error()))
+			return
+		}
+
+		sleepTime := subprocessmanager.GetDelay(elapsed, healthyProcessTime, crashCount, healthyCrashCount)
+
+		// Now we can log the process exit error, no need to escalate the error
+		if subprocessErr != nil {
+			wrapper.logger.Error("Subprocess error", zap.String("time until process restarts", sleepTime.String()), zap.String("error", subprocessErr.Error()))
+		}
+
+		time.Sleep(sleepTime)
+	}
+}
+
+// computeHealthAndCrashCount will decide whether the process is healthy, set the crashCount accordingly and shutdown the receiver if unhealthy
+func (wrapper *prometheusExecReceiver) computeHealthAndCrashCount(ctx context.Context, elapsed time.Duration, crashCount int) (int, error) {
+	if elapsed > healthyProcessTime {
+		return 1, nil
+	}
+	crashCount++
+
+	// Stop the associated receiver until process starts back up again since it is unhealthy (too little elapsed time and high crashCount)
+	// or if port is generated, to allow for new port
+	if wrapper.originalPort == 0 || crashCount > healthyCrashCount {
```

Comment:
I would recommend:

- Move the shutdown related code out of this function
- Always shutdown the receiver. The cost of doing this would presumably be relatively trivial as we don't expect processes to be crashing frequently.

With the current code, I find it hard to see if the logic around when to start the receiver matches the logic around when to shut it down. It is at least disparate.

Reply:
Moved it out of that function! However are you sure shutting down the receiver has 0 consequences? If a receiver has been alive for a while and is shut down even if "healthy" wouldn't this change values for things such as delta-type metrics? At the beginning I was shutting down every time but after discussions I thought it be better to only shut down the receiver if necessary! Let me know what you think!

Reply:
I'm not sure how it could have any impact. All the receiver does is scrape a Prometheus endpoint. As far as I'm aware it doesn't store any state, but maybe I am wrong?

What's the concern re DELTA type metrics? (Prometheus only supports CUMULATIVE as far as I'm aware)